pyln: Adds some improvements to pyln.proto.zbase32

sr-gi opened the pull request:
Before this PR, `zbase32.encode` could only be used for messages with length multiple of 5. This is because zbase32 encoding uses blocks of 5-bit words for the encoding. However, arbitrary length messages should be encodable as well by properly padding them to match the expected length.

The additions to this PR could be summarised as follows:

- Add helper functions to encode / decode arbitrary length data by padding / cutting the provided data (`_message_to_bitarray` and `_bitarray_to_message`).
- Mark any function that is not supposed to be used publicly as "private".
- Adds input checking to the public functions (`encode`, `decode`, `is_zbased32_encoded`) and raise exceptions accordingly.
- Add short descriptions to each function.
- Add tests for the whole module.

Changelog-None

cdecker commented on contrib/pyln-proto/pyln/proto/zbase32.py:
```diff
@@ -23,34 +23,69 @@
+    if filling_len:
+        # The bitarray length has to be multiple of 5. If not, it is right-filled with zeros.
+        filling = '0'*(5-filling_len)
+        barr = bitstring.ConstBitStream(bin=f"{barr.bin}{filling}")
```
Not sure we have f-string support in the lightning codebase yet. We only specified a minimum version of python in the `lightningd/plugins` repository. That being said, I'd be happy to specify one explicitly after checking with everybody.

sr-gi replied:
I can change it for `.format` otherwise.

cdecker commented on contrib/pyln-proto/pyln/proto/zbase32.py:
```diff
@@ -23,34 +23,69 @@
+    return barr
+
+
+def _bitarray_to_message(barr):
```
Type annotations here would be useful.

cdecker commented on contrib/pyln-proto/pyln/proto/zbase32.py:
```diff
@@ -34,6 +35,7 @@ def _message_to_bitarray(message):
 
 
 def _bitarray_to_message(barr):
+    """Decodes a bitarray with length multiple of 5 to a byte message (removing the trailing zeros if necessary)."""
```
Should the trimming assert that the trimmed part is all 0s?

sr-gi replied:
I was assuming the input is zbase32 encoded, in which case it must be. Given the function is supposed to be internal I didn't care about checking the input for correctness, but I can add the check and except if you think it's worth having it.

cdecker replied:
Maks sense.

cdecker commented on contrib/pyln-proto/tests/test_zbase32.py:
```diff
@@ -0,0 +1,106 @@
+
+
+def test_encode():
+    for i, message in enumerate(messages):
```
You can iterate over both arrays `messages` and `zbase_encoded_messages` using `zip(messages, zbase_encoded_messages)`. This way you can just loop like this:

```
for mess, expected in zip(messages, zbase_encoded_messages):
     got = zbase32.encode(message)
     assert(got == expected)
```

cdecker replied:
Alternatively you can also make the messages a `dict`, mapping from the raw message to the zbase32-encoded messages and then iterate over `messages.items()`.

sr-gi replied:
Yeah, indeed. That's way nicer.

cdecker reviewed (commented):
Just some minor points and some linting issues:

```
contrib/pyln-proto/pyln/proto/zbase32.py:32:22: E226 missing whitespace around arithmetic operator
contrib/pyln-proto/pyln/proto/zbase32.py:32:25: E226 missing whitespace around arithmetic operator
contrib/pyln-proto/tests/test_primitives.py:1:1: F401 'binascii.hexlify' imported but unused
contrib/pyln-proto/tests/test_primitives.py:2:1: F401 'pyln.proto.zbase32' imported but unused
Makefile:405: recipe for target 'check-python' failed
make: *** [check-python] Error 1
```

sr-gi commented on contrib/pyln-proto/pyln/proto/zbase32.py:
```diff
@@ -23,34 +23,69 @@
+
+def _bitarray_to_message(barr):
+    filling_len = len(barr) % 8
+    if filling_len:
```
Sorry, I'm too used to Python falsy checks :S. I'll change it.

sr-gi commented:
Added some fixes: 

- Fixes linting errors
- Added type annotations
- Replaced falsy checks
- Improved code style for tests

Pending:

- f-strings
- input checking for internal functions

cdecker commented:
Excellent, thanks for the fixes. There's one more failing python lint:

```
contrib/pyln-proto/pyln/proto/zbase32.py:33:27: E226 missing whitespace around arithmetic operator
```

`make check-python` will run these locally (though in the past I had differing results depending on the version of `flake8`).

sr-gi commented:
> Excellent, thanks for the fixes. There's one more failing python lint:
> 
> ```
> contrib/pyln-proto/pyln/proto/zbase32.py:33:27: E226 missing whitespace around arithmetic operator
> ```
> 
> `make check-python` will run these locally (though in the past I had differing results depending on the version of `flake8`).

Yeah, looks like we have different versions. Should be fixed now.

jsarenik commented:
The key word which should be used somewhere in the code (so that someone can `git grep` years later) is **padding**. I.e. "filling" the provided data is called "padding" them.

jsarenik commented:
So generally what i propose is `%s/filling/padding/g` (in vim syntax). Of course i may be wrong and my earlier comment may just be a false-positive caused by quick-reading the issue for the first time yesterday.

sr-gi commented:
> So generally what i propose is `%s/filling/padding/g` (in vim syntax). Of course i may be wrong and my earlier comment may just be a false-positive caused by quick-reading the issue for the first time yesterday.

Yeah, I agree, padding is the proper term. I'll rebase to change it.

cdecker commented on contrib/pyln-proto/pyln/proto/zbase32.py:
```diff
@@ -23,34 +24,76 @@
+def _message_to_bitarray(message: bytes) -> bitstring.ConstBitStream:
+    """Encodes a message as a bitarray with length multiple of 5."""
+    barr = bitstring.ConstBitStream(message)
+    padding_len = len(barr) % 5
```
Technically padding is `5 - (len(barr) % 5)`, but that's splitting hairs here :-)

sr-gi replied:
Actually yeah, it used to be called `rem` for remainder by after all the name changing now it's rather not representing its meaning anymore. Will update to fit the description.

cdecker commented on contrib/pyln-proto/pyln/proto/zbase32.py:
```diff
@@ -23,34 +24,76 @@
+    if padding_len > 0:
+        # The bitarray length has to be multiple of 5. If not, it is right-padded with zeros.
+        padding = '0' * (5 - padding_len)
+        barr = bitstring.ConstBitStream(bin=f"{barr.bin}{padding}")
```
Just one missing f-string.

cdecker reviewed (commented):
One f-string and a tiny nit. Looking good, and can merge if the f-string is addressed :-)

sr-gi commented:
Should be good now.

cdecker commented:
ACK https://github.com/ElementsProject/lightning/pull/4142/commits/883527ae3de1f2ed6193eda12961126062842760
